specs/carousel/carousel.md

```diff
@@ -0,0 +1,228 @@
+    </slot>
+    <div
+        class="carousel-items"
+        aria-live="off"
+    >
+        ${ x => 
+            return `
+                <div
+                    class="carousel-item-previous"
+                    role="group"
+                    aria-roledescription="slide"
+                    aria-label="${x.getPreviousIndex} of ${x.getItemCount}"
+                >
+                    ${x.getPreviousItem()}
+                </div>
+                <div
+                    class="carousel-item-active"
+                    role="group"
+                    aria-roledescription="slide"
+                    aria-label="${x.getActiveIndex} of ${x.getItemCount}"
+                >
+                    ${x.getActiveItem()}
+                </div>
+                <div
+                    class="carousel-item-next"
+                    role="group"
+                    aria-roledescription="slide"
+                    aria-label="${x.getNextIndex} of ${x.getItemCount}"
+                >
+                    ${x.getNextItem()}
+                </div>`
+        }
+    </div>
+    <slot name="previous-button">
+        <fast-flipper direction={FlipperDirection.previous} />
+    </slot>
+    <slot name="tab-list">Default Tab List </slot>
+</div>
+<!-- shadow dom -->
+```
+
+**Implementation**
+Note: Next and previous buttons will use the fast-flipper as default so the slots will not be required unless a custom one is passed. This example shows custom simple buttons being used.
+
+```html
+<body>
+    <fast-carousel
+        autoplay="true"
+        autoplay-interval="6500"
+        loop="true"
+        paused
+        aria-labelledby
+        aria-label
+
+        <img slot="item" />
+        <p slot="item"> Lorem ipsum... </p> 
+        <div slot="item"> some elements... </div>
+
+        <button
+            slot="next-button"
+            class="next-button"
+        >
+            <svg class="flipper-svg">
+                <path d="example path..."></path>
+            </svg>
+        </button>
+
+        <button
+            slot="previous-button"
+            class="previous-button"
+        >
+            <svg class="flipper-svg">
+                <path d="example path..."></path>
+            </svg>
+        </button>
+
+        <button
+            slot="play-toggle"
+            class="previous-button"
+        >
+            <svg class="play-svg">
+            <path d="example path..."></path>
+            </svg>
+        </button>
+
+        <div role="tab-list" class="tab-list" aria-label="A carousel of items">
+            <div tabindex="0" role="tab" class="tab-list-tab"></div>
+            <div tabindex="-1" role="tab" class="tab-list-tab"></div>
+            <div tabindex="-1" role="tab" class="tab-list-tab"></div>
+            <div tabindex="-1" role="tab" class="tab-list-tab"></div>
+            <div tabindex="-1" role="tab" class="tab-list-tab"></div>
+        </div>
+    >
+    </fast-carousel>
+</body>
+```
```

Comment:
It doesn't look like we're using the default slot. Can we put the items in the default slot?

Also, is it possible to generate the tab-list if one isn't provided? This would need to be kept in sync with the actual slotted items, but that's not too hard (I've got an idea for a library feature to help with this). These could be generated as fallback content for the tab-list slot.

Reply:
Changed to use the default.

Yeah I think we are still discussing how tab-list will be used per the W3 Carousel spec as there are some different interaction models that could be used.
I still assume it'd be best to have a default. What's your idea for the library feature to help?

Reply:
I'm thinking of having a `@slotted` decorator that would give you the list of html elements that exist in a particular slot, keeping it up to date automatically as slotted content changes:

```TypeScript
export class FastCarousel extends FastElement {
  @slotted items: HTMLElement[]; // all the html elements in the default slot 
}
````

With this in place you could repeater over `items` and generate the tab list into the default slot. Any time the slotted items change, the repeat would automatically update the tab list for you.

The `@slotted` decorator would take an optional slot name and an optional element filter for more advances scenarios. It would default to the default slot and a filter that selected all nodes.

Reply:
Another option is to do something like ref where you put a directive on your slot and it gives you a property with its content:

```
const template = html`
  <slot ${slotted('items')}></slot>
`;

Reply:
To be clear the items are the slide content, and in this template the tab-list is for the slide picker controls. Maybe I should name that something different that tab-list?

Reply:
> I'm thinking of having a `@slotted` decorator that would give you the list of html elements that exist in a particular slot, keeping it up to date automatically as slotted content changes:
> 
> ```ts
> export class FastCarousel extends FastElement {
>   @slotted items: HTMLElement[]; // all the html elements in the default slot 
> }
> ```
> 
> With this in place you could repeater over `items` and generate the tab list into the fallback content for tab list. Any time the slotted items change, the repeat would automatically update the tab list for you.
> 
> The `@slotted` decorator would take an optional slot name and an optional element filter for more advances scenarios. It would default to the default slot and a filter that selected all nodes.

I think this would be handy for what I am trying to do. I am currently having difficulty getting the nodes/html elements that are in the default slot, and then using them to render the slides, but I believe I am missing a key concept.

Reply:
You might be accessing assignedNodes to early and need to also register for the slotchanged event. What I'm proposing would do all that work for you so you just have a nice array of nodes in the end, which you could repeat over to generate other UI. I can probably get to to work on this Monday or Tuesday of next week if this sounds good.

Reply:
Sorry on the delay, this definitely sounds good.

Reply:
FYI, the new `slotted` directive is in. Right now you'll have to filter the nodes to ensure you only get element nodes, which is a pain. I'll see if I can add the filter feature in soon.